anta/reporter/md_reporter.py

```diff
@@ -0,0 +1,329 @@
+# Copyright (c) 2023-2024 Arista Networks, Inc.
+# Use of this source code is governed by the Apache License 2.0
+# that can be found in the LICENSE file.
+"""Markdown report generator for ANTA test results."""
+
+from __future__ import annotations
+
+import re
+from abc import ABC, abstractmethod
+from typing import TYPE_CHECKING, ClassVar
+
+if TYPE_CHECKING:
+    from collections.abc import Generator
+    from io import TextIOWrapper
+    from pathlib import Path
+
+    from anta.result_manager import ResultManager
+
+
+# pylint: disable=too-few-public-methods
+class MDReportGenerator:
+    """Class responsible for generating a Markdown report based on the provided `ResultManager` object.
+
+    It aggregates different report sections, each represented by a subclass of `MDReportBase`,
+    and sequentially generates their content into a markdown file.
+
+    The `generate` class method will loop over all the section subclasses and call their `generate_section` method.
+    The final report will be generated in the same order as the `sections` list of the method.
+
+    The class method also accepts an optional `only_failed_tests` flag to generate a report with only failed tests.
+
+    By default, the report will include all test results.
+    """
+
+    @classmethod
+    def generate(cls, results: ResultManager, md_filename: Path, *, only_failed_tests: bool = False) -> None:
+        """Generate and write the various sections of the markdown report.
+
+        Parameters
+        ----------
+            results: The ResultsManager instance containing all test results.
+            md_filename: The path to the markdown file to write the report into.
+            only_failed_tests: Flag to generate a report with only failed tests. Defaults to False.
+        """
+        MDReportBase.ONLY_FAILED_TESTS = only_failed_tests
+
+        with md_filename.open("w", encoding="utf-8") as mdfile:
```

Comment:
need to catch OSError here if anything goes wrong - cf #672

Reply:
Fixed in 5c449949773cb86a3c25aa336bd3d83f5150773c